OpenRA.Mods.Common/Traits/Carryable.cs

```diff
@@ -19,181 +17,89 @@ namespace OpenRA.Mods.Common.Traits
 	[Desc("Can be carried by units with the trait `Carryall`.")]
 	public class CarryableInfo : ITraitInfo, Requires<UpgradeManagerInfo>
 	{
-		[Desc("Required distance away from destination before requesting a pickup. Default is 6 cells.")]
-		public readonly WDist MinDistance = WDist.FromCells(6);
-
 		[UpgradeGrantedReference]
 		[Desc("The upgrades to grant to self while waiting or being carried.")]
 		public readonly string[] CarryableUpgrades = { };
 
-		public object Create(ActorInitializer init) { return new Carryable(init.Self, this); }
+		public readonly WDist CarryableHeight = new WDist(200);
+
+		public virtual object Create(ActorInitializer init) { return new Carryable(init.Self, this); }
 	}
 
-	public class Carryable : INotifyHarvesterAction, ICallForTransport
+	public class Carryable
 	{
 		readonly CarryableInfo info;
 		readonly Actor self;
 		readonly UpgradeManager upgradeManager;
 
-		public bool Reserved { get; private set; }
-
-		// If we're locked there isn't much we can do. We'll have to wait for the carrier to finish with us. We should not move or get new orders!
-		bool locked;
-
-		void Unlock()
-		{
-			if (!locked)
-				return;
-
-			locked = false;
-			foreach (var u in info.CarryableUpgrades)
-				upgradeManager.RevokeUpgrade(self, u, this);
-		}
-
-		void Lock()
-		{
-			if (locked)
-				return;
-
-			locked = true;
-			foreach (var u in info.CarryableUpgrades)
-				upgradeManager.GrantUpgrade(self, u, this);
-		}
-
+		public Actor Carrier { get; private set; }
+		public bool Reserved { get { return state != State.Free; } }
+		public CPos Destination { get; set; }
 		public bool WantsTransport { get; set; }
-		public CPos Destination;
-		Activity afterLandActivity;
+
+		protected enum State { Free, Reserved, Locked }
+		protected State state;
+		protected bool attached = false;
 
 		public Carryable(Actor self, CarryableInfo info)
 		{
 			this.info = info;
 			this.self = self;
 			upgradeManager = self.Trait<UpgradeManager>();
+			Destination = CPos.Zero;
```

Comment:
Using `CPos.Zero` as a magic value has caused many problems in the past.  It may be better to change this to a `CPos?` and initialize it to null.